beacon-chain/core/state/trailing_slot_state.go

```diff
@@ -0,0 +1,72 @@
+package state
+
+import (
+	"bytes"
+	"context"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+)
+
+type nextSlotCache struct {
+	sync.RWMutex
+	root  []byte
+	state *state.BeaconState
+}
+
+var (
+	nsc nextSlotCache
+	// Metrics for the validator cache.
+	nextSlotCacheHit = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "next_slot_cache_hit",
+		Help: "The total number of cache hits on the next slot state cache.",
+	})
+	nextSlotCacheMiss = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "next_slot_cache_miss",
+		Help: "The total number of cache misses on the next slot state cache.",
+	})
+)
+
+// GetNextSlotState returns the saved state if the input root matches the root in `nextSlotCache`. Returns nil otherwise.
+// This is useful to check before processing slots. With a cache hit, it will return last processed state with slot plus
+// one advancement.
+func GetNextSlotState(ctx context.Context, root []byte) (*state.BeaconState, error) {
```

Comment:
Do we want the naming convention of `Get`? I guess yes because that's what we use for caches typically, right

Reply:
Good point. Removed